src/codewhisperer/service/featureConfigProvider.ts

```diff
@@ -0,0 +1,87 @@
+import { getLogger } from '../../shared/logger'
+
+export class FeatureContext {
+    constructor(public name: string, public variation: string, public value: FeatureValue) {}
+}
+
+const testFeatureName = 'testFeature'
+const featureConfigPollIntervalInMs = 30 * 60 * 1000 // 30 mins
+
+// TODO: add real feature later
+export const featureDefinitions = new Map([
+    [testFeatureName, new FeatureContext(testFeatureName, 'CONTROL', { stringValue: 'testValue' })],
+])
+
+export class FeatureConfigProvider {
+    private featureConfigs = new Map<string, FeatureContext>()
+
+    static #instance: FeatureConfigProvider
+
+    constructor() {
+        this.fetchFeatureConfigs()
+
+        setInterval(this.fetchFeatureConfigs.bind(this), featureConfigPollIntervalInMs)
+    }
+
+    public static get instance() {
+        return (this.#instance ??= new this())
+    }
+
+    async fetchFeatureConfigs(): Promise<void> {
+        if (AuthUtil.instance.isConnectionExpired()) {
+            return
+        }
+
+        getLogger().debug('Fetching feature configs')
+        try {
+            const response = await client.listFeatureEvaluations()
+
+            // Overwrite feature configs from server response
+            response.featureEvaluations.forEach(evaluation => {
+                this.featureConfigs.set(
+                    evaluation.feature,
+                    new FeatureContext(evaluation.feature, evaluation.variation, evaluation.value)
+                )
+            })
+        } catch (e) {
+            getLogger().debug('Error when fetching feature configs', e)
+        }
+        getLogger().debug(`Current feature configs: ${this.getFeatureConfigsTelemetry()}`)
+    }
+
+    getFeatureConfigsTelemetry(): string {
+        return `{${Array.from(this.featureConfigs.entries())
+            .map(([name, context]) => `${name}: ${context.variation}`)
+            .join(', ')}}`
+    }
+
+    // TODO: for all feature variations, define a contract that can be enforced upon the implementation of
+    // the business logic.
+    // When we align on a new feature config, client-side will implement specific business logic to utilize
+    // these values by:
+    // 1) Add an entry in featureDefinitions, which is <feature_name> to <feature_context>.
+    // 2) Add a function with name `getXXX`, where XXX refers to the feature name.
+    // 3) Specify the return type: One of the return type string/boolean/Long/Double should be used here.
+    // 4) Specify the key for the `getFeatureValueForKey` helper function which is the feature name.
+    // 5) Specify the corresponding type value getter for the `FeatureValue` class. For example,
+    // if the return type is Long, then the corresponding type value getter is `longValue()`.
+    // 6) Add a test case for this feature.
+    // 7) In case `getXXX()` returns undefined, it should be treated as a default/control group.
+    getTestFeature(): string | undefined {
```

Comment:
we should make it generic instead of adding a getter for every new feature.
maybe like
```
getTestFeature<T>(FeatureContext['name']): T extends FeatureContext
```

Reply:
For me, having a different getter for each feature is slightly easier for the caller and is more controlled. Making it generic makes the caller can pass any parameters, both expected and unexpected. Plus having separate getters can also allow us to define specific return type for the specific feature, for which the contract will be very clear(thus easier to debug in general)(otherwise the caller will need to get either `stringValue` or `boolValue` which is duplicated code and hard to manage). All at the cost of adding a new getter, which I don't think it's a big deal. Anyway the feature usage is a case by case thing.